test/unit/math/mix/fun/generalized_inverse_test.cpp

```diff
@@ -0,0 +1,65 @@
+#include <test/unit/math/test_ad.hpp>
+#include <vector>
+#include <gtest/gtest.h>
+
+TEST(mathMixMatFun, ad_tests) {
+  using stan::test::expect_ad;
+
+  auto f = [](const auto& G) { return stan::math::generalized_inverse(G); };
+  auto fjit
+      = [](const auto& G) { return stan::math::generalized_inverse(G, 0.0); };
+  auto fjit2
+      = [](const auto& G) { return stan::math::generalized_inverse(G, 1e-4); };
+
+  Eigen::MatrixXd t(0, 0);
+  expect_ad(f, t);
+  expect_ad(fjit, t);
+
+  Eigen::MatrixXd u(1, 1);
+  u << 2;
+  expect_ad(f, u);
+  expect_ad(fjit, u);
+
+  Eigen::MatrixXd v(2, 3);
+  v << 1, 3, 5, 2, 4, 6;
+  expect_ad(f, v);
+  expect_ad(fjit, v);
+
+  v << 1.9, 1.3, 2.5, 0.4, 1.7, 0.1;
+  expect_ad(f, v);
+  expect_ad(fjit, v);
+
+  // issues around zero require looser tolerances for hessians
+  stan::test::ad_tolerances tols;
+  tols.hessian_hessian_ = 2.0;
+  tols.hessian_fvar_hessian_ = 2.0;
+
+  Eigen::MatrixXd w(3, 4);
+  w << 2, 3, 5, 7, 11, 13, 17, 19, 23, 25, 27, 29;
+  expect_ad(tols, f, w);
+  expect_ad(tols, fjit, w);
+
+  Eigen::MatrixXd z(2, 2);
+  z << 1, 2, 5, std::numeric_limits<double>::quiet_NaN();
+  EXPECT_NO_THROW(stan::math::generalized_inverse(z));
+  EXPECT_NO_THROW(stan::math::generalized_inverse(z, 0.0));
+
+  // autodiff throws, so following fails (throw behavior must match to pass)
+
+  Eigen::MatrixXd a(2, 2);
+  a << 1.9, 0.3, 0.3, std::numeric_limits<double>::infinity();
+  expect_ad(f, a);
+  expect_ad(fjit, a);
```

Comment:
If you want to test this works with `var_value<Matrix>` as well as `Matrix<var>` you can just add

```cpp
expect_ad_matvar(f, a);
```

so that those tests are checked as well.